### PR TITLE
[FIX] website_payment: handle type casting error for empty donation amount

### PR DESCRIPTION
**Pull request by mega-odoo**
When a user pays a donation with an empty amount value, the system passes this value as an empty string, which is leading to a type-casting error because an empty string cannot be directly converted to a numerical value like a float.

Traceback on sentry:

```
ValueError: could not convert string to float: ''
  File "odoo/http.py", line 2123, in __call__
    response = request._serve_db()
  File "odoo/http.py", line 1699, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "odoo/service/model.py", line 133, in retrying
    result = func()
  File "odoo/http.py", line 1726, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "odoo/http.py", line 1840, in dispatch
    return self.request.registry['ir.http']._dispatch(endpoint)
  File "addons/website/models/ir_http.py", line 234, in _dispatch
    response = super()._dispatch(endpoint)
  File "odoo/addons/base/models/ir_http.py", line 190, in _dispatch
    result = endpoint(**request.params)
  File "odoo/http.py", line 716, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "addons/website_payment/controllers/portal.py", line 27, in donation_pay
    kwargs['amount'] = float(kwargs.get('amount', 25))

```

To handle this situation, we should use '_cast_as_int' and '_cast_as_float' methods because these methods provide try-except blocks, effectively handling these types of errors. This ensures that when an empty amount value is encountered, the system manages the error.

sentry-4376573159

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by rdeodoo — COMMENTED** [review]
> When a user pays a donation with a blank amount value, the system passes this value as an empty string [..] This ensures that when a blank
amount value is encountered, the system gracefully manages the error, preventing
disruptions in the donation process.

1. Does that mean that you are actually going to pass `None` as amount and that the following code is in charge of raising / handling an error correctly in such a situation?
2. What do you mean by a "blank amount", how does it happens? This should not be possible, right? This commit seems to be fixing the consequence (why not, probably good) but it seems like there is a root cause to be fixed too?
3. `the system gracefully manages the error, preventing disruptions in the donation process.`  
  What does that mean? What happens in such a case? The donation fails with a proper message?

Need a bit more details, please.

**Comment by rdeodoo** [comment]
Also, your commit title is too long
image
https://github.com/odoo/odoo/pull/83508#discussion_r796442104

**Comment by mega-odoo** [comment]
> 4376573159

Passing None as the amount is causing issues, So we can do Instead of handling an error in this situation using '_cast_as_int()' and '_cast_as_float()' methods we can update a code like 'float(kwargs.get('amount') or 25)' instead of 'float(kwargs.get('amount', 25))' this is a better way. Because as per the sentry issue [4376573159](https://online.sentry.io/issues/4376573159/?project=4504605413146624&query=is%3Aunresolved+release%3Asaas~16.4+assigned%3Ame&referrer=issue-stream&sort=priority&statsPeriod=90d&stream_index=0) We can see that the 'amount' value of 'kwargs' is an 'empty string'. And this issue is not produced locally.

setry_issue

**Comment by rdeodoo** [comment]
You did not reply to question `1.` and `3.`, and you said that you couldn't reproduce it locally for question `2.`, but then:
- If this is something happening when people are messing with URL manually or stuff like that, it's generally not something we fix / support
- If this is a legit flow, we need to fix it upfront too.

Was this error seen a lot on sentry? And please do you have info to provide for `1.` and `3.`?

Having an answer for question `3.` will help reviewing this, as most likely you want to short circuit the donation process here if you are missing the currency or the amount, rather than passing `None` and hoping that the rest of the business code will work somehow correctly.

**Comment by rdeodoo** [comment]
> This issue occurs only once time in sentry thus can we close this PR?

IMO, yes, or at least I would raise a proper error here instead of doing some magic cast with default value to hope that the code will get through.

But it's for @odoo/rd-payment to give the final decision here :+1:

**Review by Demesmaeker — APPROVED** [review]
Hello @mega-odoo 

Thanks for your work. Looking a bit more here, I think we have far more important things to do with this feature. It only happens when someone manually changes the amount in the url. If we need to cover every case where the user might type random things in the url, it'll take us a very long time to manage each separate cases. What about the user that'd enter -1 as the amount, then? No error'd be thrown, but it wouldn't work either. 
Best we could do is checking that the entered amount is logical. Maybe even add a "minimum donation amount" somewhere and compare it to that (not just in the website block, where it actually works). Be it for the amount in the url, or the one in the form (when the form will work at least).
Nonetheless, this can solve this slight problem, so it's still better to have this cast than not.

tldr; this is slightly better, and we need to rework the donation feature

Have a nice day =]

File: addons/website_payment/controllers/portal.py [hunk 1]
```diff
@@ -23,8 +23,8 @@ def donation_pay(self, **kwargs):
         :raise: werkzeug.exceptions.NotFound if the access token is invalid
         """
         kwargs['is_donation'] = True
-        kwargs['currency_id'] = int(kwargs.get('currency_id', request.env.company.currency_id.id))
-        kwargs['amount'] = float(kwargs.get('amount', 25))
+        kwargs['currency_id'] = self._cast_as_int(kwargs.get('currency_id', request.env.company.currency_id.id))
+        kwargs['amount'] = self._cast_as_float(kwargs.get('amount', 25))
```
[the remark this answers is not in the document]
Comment (mega-odoo) [reply addons/website_payment/controllers/portal.py]: Hi, it's a better way to handle an error. So changes have been made according to your suggestion, Can you please look into it?

Thanks.

**Review by Demesmaeker — APPROVED** [review]
Thanks for the changes ^^

cc: @AntoineVDV

**Review by AntoineVDV — APPROVED** [review]
@robodoo r+
